File: apps/pipeline/src/pipeline/drift.py
"""Detect corpus drift and publish explicitly reviewed immutable versions."""

from datetime import UTC, datetime

from pipeline.corpus import compile_corpus, content_digest


SIGNALS = (
    "input",
    "tool",
    "evaluator",
    "acceptance",
    "model",
    "cost",
    "latency",
    "retry",
    "trajectory",
)


def _run_map(bundle):
    return {run["id"]: run for run in bundle["runs"]}


def _case_map(cases):
    return {case["case_id"]: case for case in cases}


def _different(left, right, key):
    return (
        left.get(key) is not None and right.get(key) is not None and left.get(key) != right.get(key)
    )


def _run_signals(parent_run, candidate_run):
    signals = []
    if _different(parent_run, candidate_run, "prompt"):
        signals.append("input")
    if _different(parent_run, candidate_run, "tool_calls"):
        signals.append("tool")
    if _different(parent_run, candidate_run, "evaluator") or _different(
        parent_run, candidate_run, "evaluator_version"
    ):
        signals.append("evaluator")
    if _different(parent_run, candidate_run, "success"):
        signals.append("acceptance")
    if _different(parent_run, candidate_run, "model"):
        signals.append("model")
    if _relative_change(parent_run.get("cost_usd"), candidate_run.get("cost_usd")) > 0.2:
        signals.append("cost")
    if (
        _relative_change(
            parent_run.get("duration_ms", parent_run.get("latency_ms")),
            candidate_run.get("duration_ms", candidate_run.get("latency_ms")),
        )
        > 0.2
    ):
        signals.append("latency")
    if _different(parent_run, candidate_run, "retry_count"):
        signals.append("retry")
    if _different(parent_run, candidate_run, "trajectory"):
        signals.append("trajectory")
    return signals


def _relative_change(left, right):
    if not isinstance(left, (int, float)) or not isinstance(right, (int, float)) or left == 0:
        return 0
    return abs(right - left) / abs(left)


def _case_signal(parent_case, candidate_case):
    comparable = [
        "pipeline_family",
        "workload_label",
        "risk",
        "required_evidence",
        "checks",
        "labels",
    ]
    return (
        "evaluator"
        if any(parent_case.get(key) != candidate_case.get(key) for key in comparable)
        else None
    )


def _change(action, case_id, signal, detail):
    return {
        "action": action,
        "case_id": case_id,
        "signal": signal,
        "detail": detail,
    }


def _proposal_body(parent_manifest, candidate_definition, changes, signals, exposed, replacements):
    return {
        "version": "1",
        "parent_corpus_version_id": parent_manifest["content_digest"],
        "candidate_definition_digest": content_digest(candidate_definition),
        "status": "proposed",
        "proposed_changes": changes,
        "signals": sorted(set(signals)) or ["none"],
        "exposed_holdout_case_ids": sorted(exposed),
        "replacement_holdout_case_ids": sorted(replacements),
        "approval": None,
    }


def detect_drift(parent_manifest, parent_bundle, candidate_bundle, candidate_definition):
    parent_cases = _case_map(parent_manifest["cases"])
    candidate_cases = _case_map(candidate_definition["cases"])
    parent_runs = _run_map(parent_bundle)
    candidate_runs = _run_map(candidate_bundle)
    changes = []
    signals = []

    for case_id in sorted(set(candidate_cases) - set(parent_cases)):
        changes.append(_change("add", case_id, "none", "case is new in the candidate corpus"))
    for case_id in sorted(set(parent_cases) - set(candidate_cases)):
        changes.append(
            _change("retire", case_id, "none", "case is absent from the candidate corpus")
        )

    for case_id in sorted(set(parent_cases) & set(candidate_cases)):
        parent_case = parent_cases[case_id]
        candidate_case = candidate_cases[case_id]
        case_signal = _case_signal(parent_case, candidate_case)
        if case_signal:
            action = "red-team" if candidate_case["risk"] == "high" else "relabel"
            changes.append(
                _change(action, case_id, case_signal, "reviewed case definition changed")
            )
            signals.append(case_signal)
        parent_run = parent_runs.get(parent_case["source_run_id"])
        candidate_run = candidate_runs.get(candidate_case["source_run_id"])
        if parent_run and candidate_run:
            for signal in _run_signals(parent_run, candidate_run):
                action = "red-team" if candidate_case["risk"] == "high" else "relabel"
                changes.append(_change(action, case_id, signal, f"source run shows {signal} drift"))
                signals.append(signal)

    exposed = sorted(
        case_id
        for case_id, parent_case in parent_cases.items()
        if parent_case["split"] == "holdout"
        and case_id in candidate_cases
        and candidate_cases[case_id]["split"] == "working"
    )
    replacements = sorted(
        case_id
        for case_id, candidate_case in candidate_cases.items()
        if candidate_case["split"] == "holdout" and case_id not in parent_cases
    )
    if exposed:
        signals.append("acceptance")
        changes.append(
            _change(
                "red-team",
                exposed[0],
                "acceptance",
                "holdout case was exposed and needs replacement coverage",
            )
        )
    if not changes:
        changes = [_change("no-change", "corpus", "none", "no material drift detected")]
    body = _proposal_body(
        parent_manifest,
        candidate_definition,
        changes,
        signals,
        exposed,
        replacements,
    )
    return {**body, "proposal_id": content_digest(body)}


def approve_drift(proposal, actor, reason=None, timestamp=None):
    if proposal["status"] != "proposed":
        raise ValueError("only proposed drift may be approved")
    body = dict(proposal)
    proposal_id = body.pop("proposal_id")
    body.pop("status")
    body.pop("approval")
    if content_digest({**body, "status": "proposed", "approval": None}) != proposal_id:
        raise ValueError("drift proposal digest does not match its contents")
    approved = {
        **proposal,
        "status": "approved",
        "approval": {
            "actor": actor,
            "timestamp": timestamp or datetime.now(UTC).isoformat(),
            "reason": reason,
        },
    }
    return approved


def publish_corpus(parent_manifest, candidate_bundle, candidate_definition, proposal):
    if proposal["status"] != "approved" or proposal.get("approval") is None:
        raise ValueError("corpus publication requires an approved drift proposal")
    proposal_body = dict(proposal)
    proposal_id = proposal_body.pop("proposal_id")
    proposal_body.pop("status")
    proposal_body.pop("approval")
    if content_digest({**proposal_body, "status": "proposed", "approval": None}) != proposal_id:
        raise ValueError("drift proposal digest does not match its contents")
    if proposal["parent_corpus_version_id"] != parent_manifest["content_digest"]:
        raise ValueError("drift proposal parent does not match the current corpus")
    if proposal["candidate_definition_digest"] != content_digest(candidate_definition):
        raise ValueError("candidate definition does not match drift proposal")
    if candidate_definition["parent_version"] != parent_manifest["content_digest"]:
        raise ValueError("candidate definition must name the current corpus as its parent")
    exposed = set(proposal["exposed_holdout_case_ids"])
    replacements = set(proposal["replacement_holdout_case_ids"])
    candidate_case_ids = {case["case_id"] for case in candidate_definition["cases"]}
    if not exposed.issubset(candidate_case_ids):
        raise ValueError("drift proposal references missing exposed holdout cases")
    if len(replacements) < len(exposed):
        raise ValueError("every exposed holdout requires a replacement holdout case")
    manifest, benchmark_cases = compile_corpus(candidate_bundle, candidate_definition)
    if manifest["content_digest"] == parent_manifest["content_digest"]:
        raise ValueError("published corpus must create a new immutable version")
    return manifest, benchmark_cases

File: harness/packages/kernel/src/drift.ts
import {
  canonicalJson,
  computeRunSpecDigest,
  type JsonValue,
} from "@rightmodeler/core";

export const DRIFT_SIGNALS = [
  "input",
  "tool",
  "evaluator",
  "acceptance",
  "model",
  "cost",
  "latency",
  "retry",
  "trajectory",
  "none",
] as const;

export type DriftSignal = (typeof DRIFT_SIGNALS)[number];
export type DriftChangeAction =
  "add" | "relabel" | "retire" | "red-team" | "no-change";
export type PipelineFamily =
  "reference-freeform" | "structured-check" | "tool-trajectory" | "repo-fix";

export interface HistoricalRun {
  readonly id: string;
  readonly prompt: string;
  readonly model: string;
  readonly tool_calls?: JsonValue;
  readonly evaluator?: JsonValue;
  readonly evaluator_version?: JsonValue;
  readonly success?: boolean | null;
  readonly cost_usd?: number | null;
  readonly duration_ms?: number | null;
  readonly latency_ms?: number | null;
  readonly retry_count?: number | null;
  readonly trajectory?: JsonValue;
  readonly not_evaluated?: readonly DriftNotEvaluated[];
}

export interface DriftNotEvaluated {
  readonly signal: DriftSignal;
  readonly dimension: string;
  readonly reason: string;
}

export interface HistoricalRunBundle {
  readonly version: string;
  readonly bundle_id: string;
  readonly runs: readonly HistoricalRun[];
}

export interface CorpusCaseLabels {
  readonly recommendation: "safe" | "unsafe" | "not_applicable";
  readonly required_abstention: boolean;
}

export interface CorpusCaseDefinition {
  readonly case_id: string;
  readonly source_run_id: string;
  readonly pipeline_family: PipelineFamily;
  readonly workload_label: string;
  readonly split: "working" | "holdout";
  readonly risk: "normal" | "high";
  readonly required_evidence:
    "deterministic" | "reference" | "trajectory" | "abstain";
  readonly checks: Readonly<Record<string, JsonValue>>;
  readonly labels?: CorpusCaseLabels;
}

export interface CorpusDefinition {
  readonly version: string;
  readonly corpus_id: string;
  readonly parent_version: string | null;
  readonly cases: readonly CorpusCaseDefinition[];
}

export interface CorpusManifest {
  readonly version: string;
  readonly corpus_id: string;
  readonly corpus_version: string;
  readonly parent_version: string | null;
  readonly source_bundle_id: string;
  readonly content_digest: string;
  readonly cases: readonly CorpusCaseDefinition[];
}

export interface BenchmarkCases {
  readonly version: string;
  readonly corpus_version_id: string;
  readonly source_bundle_id: string;
  readonly cases: readonly CorpusCaseDefinition[];
}

export interface DriftChange {
  readonly action: DriftChangeAction;
  readonly case_id: string;
  readonly signal: DriftSignal;
  readonly detail: string;
}

export interface DriftApproval {
  readonly actor: string;
  readonly timestamp: string;
  readonly reason: string | null;
}

interface CorpusDriftProposalBase {
  readonly version: string;
  readonly proposal_id: string;
  readonly parent_corpus_version_id: string;
  readonly candidate_definition_digest: string;
  readonly proposed_changes: readonly DriftChange[];
  readonly signals: readonly DriftSignal[];
  readonly exposed_holdout_case_ids: readonly string[];
  readonly replacement_holdout_case_ids: readonly string[];
}

export type ProposedCorpusDriftProposal = CorpusDriftProposalBase & {
  readonly status: "proposed";
  readonly approval: null;
};

export type ApprovedCorpusDriftProposal = CorpusDriftProposalBase & {
  readonly status: "approved";
  readonly approval: DriftApproval;
};

export type PublishedCorpusDriftProposal = CorpusDriftProposalBase & {
  readonly status: "published";
  readonly approval: DriftApproval;
};

export type RejectedCorpusDriftProposal = CorpusDriftProposalBase & {
  readonly status: "rejected";
  readonly approval: DriftApproval | null;
};

export type CorpusDriftProposal =
  | ProposedCorpusDriftProposal
  | ApprovedCorpusDriftProposal
  | PublishedCorpusDriftProposal
  | RejectedCorpusDriftProposal;

export interface PublishDriftResult {
  readonly proposal: PublishedCorpusDriftProposal;
  readonly manifest: CorpusManifest;
  readonly benchmark_cases: BenchmarkCases;
}

const pipelineFamilies = new Set<PipelineFamily>([
  "reference-freeform",
  "structured-check",
  "tool-trajectory",
  "repo-fix",
]);

export function detectDrift(
  parentManifest: CorpusManifest,
  parentBundle: HistoricalRunBundle,
  candidateBundle: HistoricalRunBundle,
  candidateDefinition: CorpusDefinition,
): ProposedCorpusDriftProposal {
  const parentCases = caseMapBySourceRunId(parentManifest.cases);
  const candidateCases = caseMapBySourceRunId(candidateDefinition.cases);
  const parentRuns = runMap(parentBundle);
  const candidateRuns = runMap(candidateBundle);
  let changes: DriftChange[] = [];
  const signals: DriftSignal[] = [];

  for (const sourceRunId of difference(candidateCases, parentCases)) {
    const caseId = candidateCases.get(sourceRunId)!.case_id;
    changes.push(
      change("add", caseId, "none", "case is new in the candidate corpus"),
    );
  }
  for (const sourceRunId of difference(parentCases, candidateCases)) {
    const caseId = parentCases.get(sourceRunId)!.case_id;
    changes.push(
      change(
        "retire",
        caseId,
        "none",
        "case is absent from the candidate corpus",
      ),
    );
  }

  const notEvaluated = new Map<string, DriftChange>();
  for (const sourceRunId of intersection(parentCases, candidateCases)) {
    const parentCase = parentCases.get(sourceRunId)!;
    const candidateCase = candidateCases.get(sourceRunId)!;
    const caseId = candidateCase.case_id;
    if (caseDefinitionChanged(parentCase, candidateCase)) {
      changes.push(
        change(
          candidateCase.risk === "high" ? "red-team" : "relabel",
          caseId,
          "evaluator",
          "reviewed case definition changed",
        ),
      );
      signals.push("evaluator");
    }

    const parentRun = parentRuns.get(parentCase.source_run_id);
    const candidateRun = candidateRuns.get(candidateCase.source_run_id);
    if (parentRun !== undefined && candidateRun !== undefined) {
      for (const signal of runSignals(parentRun, candidateRun)) {
        changes.push(
          change(
            candidateCase.risk === "high" ? "red-team" : "relabel",
            caseId,
            signal,
            `source run shows ${signal} drift`,
          ),
        );
        signals.push(signal);
      }
      for (const item of [
        ...(parentRun.not_evaluated ?? []),
        ...(candidateRun.not_evaluated ?? []),
      ]) {
        const detail = `${item.dimension} drift not evaluated: ${item.reason}`;
        notEvaluated.set(
          `${item.signal}\0${item.dimension}\0${item.reason}`,
          change("no-change", caseId, item.signal, detail),
        );
      }
    }
  }

  const exposed = [...parentCases.entries()]
    .flatMap(([sourceRunId, parentCase]) =>
      parentCase.split === "holdout" &&
      candidateCases.get(sourceRunId)?.split === "working"
        ? [candidateCases.get(sourceRunId)!.case_id]
        : [],
    )
    .sort(compareText);
  const replacements = [...candidateCases.entries()]
    .flatMap(([sourceRunId, candidateCase]) =>
      candidateCase.split === "holdout" && !parentCases.has(sourceRunId)
        ? [candidateCase.case_id]
        : [],
    )
    .sort(compareText);
  if (exposed.length > 0) {
    signals.push("acceptance");
    changes.push(
      change(
        "red-team",
        exposed[0]!,
        "acceptance",
        "holdout case was exposed and needs replacement coverage",
      ),
    );
  }
  changes.push(
    ...[...notEvaluated.values()].sort(
      (left, right) =>
        compareText(left.detail, right.detail) ||
        compareText(left.case_id, right.case_id),
    ),
  );
  if (changes.length === 0) {
    changes = [
      change("no-change", "corpus", "none", "no material drift detected"),
    ];
  }

  const body = {
    version: "1",
    parent_corpus_version_id: parentManifest.content_digest,
    candidate_definition_digest: contentDigest(
      corpusDefinitionValue(candidateDefinition),
    ),
    status: "proposed" as const,
    proposed_changes: changes,
    signals:
      signals.length === 0
        ? (["none"] as const)
        : [...new Set(signals)].sort(compareText),
    exposed_holdout_case_ids: exposed,
    replacement_holdout_case_ids: replacements,
    approval: null,
  };
  return {
    ...body,
    proposal_id: contentDigest(proposalBodyValue(body)),
  };
}

export function approveDrift(
  proposal: CorpusDriftProposal,
  actor: string,
  reason: string | null = null,
  timestamp = new Date().toISOString(),
): ApprovedCorpusDriftProposal {
  if (proposal.status !== "proposed") {
    throw new Error("only proposed drift may be approved");
  }
  assertDriftProposalIntegrity(proposal);
  return {
    ...proposal,
    status: "approved",
    approval: { actor, timestamp, reason },
  };
}

export function publishDrift(
  parentManifest: CorpusManifest,
  candidateBundle: HistoricalRunBundle,
  candidateDefinition: CorpusDefinition,
  proposal: CorpusDriftProposal,
): PublishDriftResult {
  if (proposal.status !== "approved" || proposal.approval === null) {
    throw new Error("corpus publication requires an approved drift proposal");
  }
  assertDriftProposalIntegrity(proposal);
  if (proposal.parent_corpus_version_id !== parentManifest.content_digest) {
    throw new Error("drift proposal parent does not match the current corpus");
  }
  let candidateDefinitionDigest: string;
  try {
    candidateDefinitionDigest = contentDigest(
      corpusDefinitionValue(candidateDefinition),
    );
  } catch {
    throw new Error("candidate definition does not match drift proposal");
  }
  if (proposal.candidate_definition_digest !== candidateDefinitionDigest) {
    throw new Error("candidate definition does not match drift proposal");
  }
  if (candidateDefinition.parent_version !== parentManifest.content_digest) {
    throw new Error(
      "candidate definition must name the current corpus as its parent",
    );
  }

  const candidateCaseIds = new Set(
    candidateDefinition.cases.map(({ case_id }) => case_id),
  );
  if (
    proposal.exposed_holdout_case_ids.some(
      (caseId) => !candidateCaseIds.has(caseId),
    )
  ) {
    throw new Error("drift proposal references missing exposed holdout cases");
  }
  const exposed = new Set(proposal.exposed_holdout_case_ids);
  const replacements = new Set(proposal.replacement_holdout_case_ids);
  if (replacements.size < exposed.size) {
    throw new Error(
      "every exposed holdout requires a replacement holdout case",
    );
  }

  const { manifest, benchmarkCases } = compileCorpus(
    candidateBundle,
    candidateDefinition,
  );
  return {
    proposal: {
      ...proposal,
      status: "published",
      approval: proposal.approval,
    },
    manifest,
    benchmark_cases: benchmarkCases,
  };
}

function runMap(bundle: HistoricalRunBundle): Map<string, HistoricalRun> {
  return new Map(bundle.runs.map((run) => [run.id, run]));
}

function caseMapBySourceRunId(
  cases: readonly CorpusCaseDefinition[],
): Map<string, CorpusCaseDefinition> {
  return new Map(
    cases.map((corpusCase) => [corpusCase.source_run_id, corpusCase]),
  );
}

function difference(
  left: ReadonlyMap<string, unknown>,
  right: ReadonlyMap<string, unknown>,
): string[] {
  return [...left.keys()].filter((key) => !right.has(key)).sort(compareText);
}

function intersection(
  left: ReadonlyMap<string, unknown>,
  right: ReadonlyMap<string, unknown>,
): string[] {
  return [...left.keys()].filter((key) => right.has(key)).sort(compareText);
}

function different(left: JsonValue | undefined, right: JsonValue | undefined) {
  return (
    left !== undefined &&
    left !== null &&
    right !== undefined &&
    right !== null &&
    canonicalJson(left) !== canonicalJson(right)
  );
}

function runSignals(
  parentRun: HistoricalRun,
  candidateRun: HistoricalRun,
): DriftSignal[] {
  const signals: DriftSignal[] = [];
  if (different(parentRun.prompt, candidateRun.prompt)) signals.push("input");
  if (different(parentRun.tool_calls, candidateRun.tool_calls)) {
    signals.push("tool");
  }
  if (
    different(parentRun.evaluator, candidateRun.evaluator) ||
    different(parentRun.evaluator_version, candidateRun.evaluator_version)
  ) {
    signals.push("evaluator");
  }
  if (different(parentRun.success, candidateRun.success)) {
    signals.push("acceptance");
  }
  if (different(parentRun.model, candidateRun.model)) signals.push("model");
  if (relativeChange(parentRun.cost_usd, candidateRun.cost_usd) > 0.2) {
    signals.push("cost");
  }
  if (relativeChange(runLatency(parentRun), runLatency(candidateRun)) > 0.2) {
    signals.push("latency");
  }
  if (different(parentRun.retry_count, candidateRun.retry_count)) {
    signals.push("retry");
  }
  if (different(parentRun.trajectory, candidateRun.trajectory)) {
    signals.push("trajectory");
  }
  return signals;
}

function relativeChange(
  left: number | null | undefined,
  right: number | null | undefined,
): number {
  if (typeof left !== "number" || typeof right !== "number" || left === 0) {
    return 0;
  }
  return Math.abs(right - left) / Math.abs(left);
}

function runLatency(run: HistoricalRun): number | null | undefined {
  return run.duration_ms === undefined ? run.latency_ms : run.duration_ms;
}

function caseDefinitionChanged(
  parentCase: CorpusCaseDefinition,
  candidateCase: CorpusCaseDefinition,
): boolean {
  return (
    !sameValue(parentCase.pipeline_family, candidateCase.pipeline_family) ||
    !sameValue(parentCase.workload_label, candidateCase.workload_label) ||
    !sameValue(parentCase.risk, candidateCase.risk) ||
    !sameValue(parentCase.required_evidence, candidateCase.required_evidence) ||
    !sameValue(parentCase.checks, candidateCase.checks) ||
    !sameValue(
      corpusLabelsValue(parentCase.labels),
      corpusLabelsValue(candidateCase.labels),
    )
  );
}

function sameValue(
  left: JsonValue | undefined,
  right: JsonValue | undefined,
): boolean {
  if (left === undefined || right === undefined) return left === right;
  return canonicalJson(left) === canonicalJson(right);
}

function change(
  action: DriftChangeAction,
  case_id: string,
  signal: DriftSignal,
  detail: string,
): DriftChange {
  return { action, case_id, signal, detail };
}

export function assertDriftProposalIntegrity(
  proposal: CorpusDriftProposal,
): void {
  let digest: string;
  try {
    digest = contentDigest(proposalBodyValue(proposal));
  } catch {
    throw new Error("drift proposal digest does not match its contents");
  }
  if (digest !== proposal.proposal_id) {
    throw new Error("drift proposal digest does not match its contents");
  }
}

export function assertCorpusManifestIntegrity(manifest: CorpusManifest): void {
  let digest: string;
  try {
    digest = contentDigest(manifestBodyValue(manifest));
  } catch {
    throw new Error("corpus manifest digest does not match its contents");
  }
  if (digest !== manifest.content_digest) {
    throw new Error("corpus manifest digest does not match its contents");
  }
}

function contentDigest(value: JsonValue): string {
  return `sha256:${computeRunSpecDigest(value)}`;
}

function proposalBodyValue(
  proposal: Omit<CorpusDriftProposal, "proposal_id"> | CorpusDriftProposal,
): JsonValue {
  return {
    version: proposal.version,
    parent_corpus_version_id: proposal.parent_corpus_version_id,
    candidate_definition_digest: proposal.candidate_definition_digest,
    status: "proposed",
    proposed_changes: proposal.proposed_changes.map((item) => ({ ...item })),
    signals: [...proposal.signals],
    exposed_holdout_case_ids: [...proposal.exposed_holdout_case_ids],
    replacement_holdout_case_ids: [...proposal.replacement_holdout_case_ids],
    approval: null,
  };
}

function corpusCaseValue(corpusCase: CorpusCaseDefinition): JsonValue {
  return {
    case_id: corpusCase.case_id,
    source_run_id: corpusCase.source_run_id,
    pipeline_family: corpusCase.pipeline_family,
    workload_label: corpusCase.workload_label,
    split: corpusCase.split,
    risk: corpusCase.risk,
    required_evidence: corpusCase.required_evidence,
    checks: { ...corpusCase.checks },
    ...(corpusCase.labels === undefined
      ? {}
      : { labels: { ...corpusCase.labels } }),
  };
}

function corpusLabelsValue(
  labels: CorpusCaseLabels | undefined,
): JsonValue | undefined {
  return labels === undefined ? undefined : { ...labels };
}

function corpusDefinitionValue(definition: CorpusDefinition): JsonValue {
  return {
    version: definition.version,
    corpus_id: definition.corpus_id,
    parent_version: definition.parent_version,
    cases: definition.cases.map(corpusCaseValue),
  };
}

function manifestBodyValue(
  manifest: Omit<CorpusManifest, "content_digest"> | CorpusManifest,
): JsonValue {
  return {
    version: manifest.version,
    corpus_id: manifest.corpus_id,
    corpus_version: manifest.corpus_version,
    parent_version: manifest.parent_version,
    source_bundle_id: manifest.source_bundle_id,
    cases: manifest.cases.map(corpusCaseValue),
  };
}

function compileCorpus(
  bundle: HistoricalRunBundle,
  definition: CorpusDefinition,
): { manifest: CorpusManifest; benchmarkCases: BenchmarkCases } {
  const runs = runMap(bundle);
  const cases: CorpusCaseDefinition[] = [];
  const seenCaseIds = new Set<string>();

  for (const corpusCase of definition.cases) {
    if (seenCaseIds.has(corpusCase.case_id)) {
      throw new Error(`duplicate corpus case: ${corpusCase.case_id}`);
    }
    seenCaseIds.add(corpusCase.case_id);
    const sourceRun = runs.get(corpusCase.source_run_id);
    if (sourceRun === undefined) {
      throw new Error(`source run not found: ${corpusCase.source_run_id}`);
    }
    if (sourceRun.success !== true) {
      throw new Error(
        `source run is not accepted: ${corpusCase.source_run_id}`,
      );
    }
    if (!pipelineFamilies.has(corpusCase.pipeline_family)) {
      throw new Error(
        `unsupported pipeline family: ${corpusCase.pipeline_family}`,
      );
    }
    cases.push(cloneCorpusCase(corpusCase));
  }
  cases.sort((left, right) => compareText(left.case_id, right.case_id));

  const manifestWithoutDigest = {
    version: definition.version,
    corpus_id: definition.corpus_id,
    corpus_version: definition.version,
    parent_version: definition.parent_version,
    source_bundle_id: bundle.bundle_id,
    cases,
  };
  const digest = contentDigest(manifestBodyValue(manifestWithoutDigest));
  const manifest: CorpusManifest = {
    ...manifestWithoutDigest,
    content_digest: digest,
  };
  return {
    manifest,
    benchmarkCases: {
      version: definition.version,
      corpus_version_id: digest,
      source_bundle_id: bundle.bundle_id,
      cases,
    },
  };
}

function cloneCorpusCase(
  corpusCase: CorpusCaseDefinition,
): CorpusCaseDefinition {
  return {
    ...corpusCase,
    checks: { ...corpusCase.checks },
    ...(corpusCase.labels === undefined
      ? {}
      : { labels: { ...corpusCase.labels } }),
  };
}

function compareText(left: string, right: string): number {
  return left < right ? -1 : left > right ? 1 : 0;
}
